tt_llk_common/llk_san_types.h

```diff
@@ -0,0 +1,280 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+#include "llk_defs.h"
+
+namespace llk_san
+{
+
+struct dontcare_t
+{
+    constexpr dontcare_t() = default;
+};
+
+struct indeterminate_t
+{
+    constexpr indeterminate_t() = default;
+};
+
+static constexpr dontcare_t DONTCARE           = dontcare_t();
+static constexpr indeterminate_t INDETERMINATE = indeterminate_t();
+
+enum class state_type_t : uint8_t
+{
+    determinate,
+    indeterminate,
+    dontcare
+};
+
+template <typename T>
+class state_t;
+
+template <typename T>
+struct is_state_t : std::false_type
+{
+};
+
+template <typename T>
+struct is_state_t<state_t<T>> : std::true_type
+{
+};
+
+template <typename T>
+inline constexpr bool is_state_t_v = is_state_t<T>::value;
+
+template <typename T>
+class state_t
+{
+private:
+    T underlying;
+    state_type_t state_type;
+
+public:
+    // CONSTRUCTION
+    // Default to INDETERMINATE state because hardware is not initialized
+    state_t() : underlying {}, state_type(state_type_t::indeterminate)
+    {
+    }
+
+    state_t(const state_t&) = default;
+    state_t(state_t&&)      = default;
+
+    // CONVERSION
+    // - llk_san::DONTCARE      -> state_t with state_type == dontcare
+    // - llk_san::INDETERMINATE -> state_t with state_type == indeterminate
+    // - other                  -> state_t with state_type == determinate (storing the value)
+
+    // Constructor for DONTCARE
+    constexpr state_t(const dontcare_t&) : underlying {}, state_type(state_type_t::dontcare)
+    {
+    }
+
+    // Constructor for INDETERMINATE
+    constexpr state_t(const indeterminate_t&) : underlying {}, state_type(state_type_t::indeterminate)
+    {
+    }
+
+    // Constructor for DETERMINATE value
+    template <
+        typename U,
+        typename = std::enable_if_t<
+            !is_state_t_v<std::decay_t<U>> && !std::is_same_v<std::decay_t<U>, dontcare_t> && !std::is_same_v<std::decay_t<U>, indeterminate_t>>>
+    constexpr state_t(U&& value) : underlying(std::forward<U>(value)), state_type(state_type_t::determinate)
+    {
+    }
+
+    // ASSIGNMENT
+    // if RHS of assignment is state_type_t::dontcare, noop (stays old value)
+    // otherwise take the state_type and underlying of RHS
+
+    // Template operators contain the actual logic
+    template <typename U>
+    state_t& operator=(const state_t<U>& rhs)
+    {
+        if (rhs.state_type == state_type_t::dontcare)
+        {
+            return *this;
+        }
+
+        state_type = rhs.state_type;
+        underlying = rhs.underlying;
+
+        return *this;
+    }
+
+    template <typename U>
+    state_t& operator=(state_t<U>&& rhs)
```

Comment:
```suggestion
    template <typename U, typename = std::enable_if_t<std::is_same_v<T, U> || std::is_assignable_v<T&, U&&>>>
    state_t& operator=(state_t<U>&& rhs) noexcept(std::is_nothrow_assignable_v<T&, U&&>)
```
Ditto. I hope I haven't made a typo somewhere...